in_toto/models/metadata.py

```diff
@@ -23,20 +23,86 @@
 """
 
 import attr
+import copy
 import json
+from typing import Optional, List, Union
 
 import securesystemslib.keys
 import securesystemslib.formats
 import securesystemslib.exceptions
 import securesystemslib.gpg.functions
+from securesystemslib.key import Key
+from securesystemslib.metadata import Envelope as SSlibEnvelope
+from securesystemslib.signer import GPGSignature, Signature, Signer
+from securesystemslib.serialization import (BaseDeserializer, BaseSerializer,
+    JSONDeserializer, JSONSerializer, SerializationMixin)
+from in_toto.exceptions import InvalidMetadata
 
 from in_toto.models.common import ValidationMixin
 from in_toto.models.link import Link
 from in_toto.models.layout import Layout
-from in_toto.exceptions import SignatureVerificationError
+
+
+class AnyMetadataDeserializer(JSONDeserializer):
+  """Deserialize bytes into Metablock / DSSE Envelope."""
+
+  def deserialize(self, raw_data: bytes) -> Union["Envelope", "Metablock"]:
+    """Deserialize JSON bytes into Metablock or Envelope as per ITE-5. """
+
+    data = super().deserialize(raw_data)
+
+    if "payload" in data:
+      if data.get("payloadType") == "application/vnd.in-toto+json":
+        return Envelope.from_dict(data)
+
+    elif "signed" in data:
+      return Metablock.from_dict(data)
+
+    raise InvalidMetadata
+
+
+class AnyMetadata(SerializationMixin):
+  """A Metadata abstraction between DSSE Envelope and Metablock."""
+
+  @staticmethod
+  def default_deserializer() -> BaseDeserializer:
+    return AnyMetadataDeserializer()
+
+  @staticmethod
+  def default_serializer() -> BaseSerializer:
+    return JSONSerializer()
+
+
+class PayloadDeserializer(JSONDeserializer):
+  """Deserialize JSON bytes into Link or Layout."""
+
+  def deserialize(self, raw_data: bytes) -> Union[Link, Layout]:
+    data = super().deserialize(raw_data)
+    _type = data.get("_type")
+    if _type == "link":
+      return Link.read(data)
+    if _type == "layout":
+      return Layout.read(data)
+
+    raise securesystemslib.exceptions.DeserializationError(
+      f"Invalid payload type {_type}, must be `link` or `layout`"
+    )
+
+
+class Envelope(AnyMetadata, SSlibEnvelope):
+  """in-toto copy of DSSE Envelope."""
```

Comment:
Not really a copy. What about
```suggestion
  """DSSE Envelope for in-toto payloads."""
```